src/System.Private.CoreLib/shared/System/Diagnostics/Tracing/CounterGroup.cs

```diff
@@ -205,48 +239,52 @@ private void OnTimer(object? state)
                         counter.WritePayload((float)elapsed.TotalSeconds, _pollingIntervalInMilliseconds);
                     }
                     _timeStampSinceCollectionStarted = now;
-                }
-                else
-                {
-                    DisposeTimer();
+
+                    do {
+                        nextPollingTimeStamp += new TimeSpan(0, 0, 0, 0, _pollingIntervalInMilliseconds);
+                    } while(nextPollingTimeStamp <= now);
                 }
             }
         }
 
-        #region PCL timer hack
-
-#if ES_BUILD_PCL
-        internal delegate void TimerCallback(object state);
 
-        internal sealed class Timer : CancellationTokenSource, IDisposable
-        {
-            private int _period;
-            private TimerCallback _callback;
-            private object _state;
 
-            internal Timer(TimerCallback callback, object state, int dueTime, int period)
-            {
-                _callback = callback;
-                _state = state;
-                _period = period;
-                Schedule(dueTime);
-            }
+        private static Thread? s_pollingThread;
+        private static int s_sleepDurationInMilliseconds;
+        private static ManualResetEvent? s_pollingThreadEvent;
+        private static List<CounterGroup>? s_counterGroupEnabledList;
 
-            private void Schedule(int dueTime)
-            {
-                Task.Delay(dueTime, Token).ContinueWith(OnTimer, null, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
-            }
-
-            private void OnTimer(Task t, object s)
+        private static void PollForValues()
+        {
+            while (true)
             {
-                Schedule(_period);
-                _callback(_state);
+                while (s_counterGroupEnabledList?.Count > 0)
+                {
+                    lock (s_counterGroupLock)
+                    {
+                        foreach (CounterGroup counterGroup in s_counterGroupEnabledList)
+                        {
+                            if (counterGroup.nextPollingTimeStamp < DateTime.UtcNow)
```

Comment:
The logic here doesn't handle tiny differences in time well. Imagine:
nextPollingTime = 1.0 sec
DateTime.UtcNow = 0.0007 sec
millisecondsTillNextPoll = (int)(1.0-0.007).TotalMilliseconds = 999
sleep(999) happens, thread wakes up and DateTime.UtcNow = 0.9997
counterGroup.nextPollingTimeStamp (1.000) < DateTime.UtcNow (0.9997) evaluates false, OnTimer() doesn't run
millisecondsTillNextPoll = (int)(1.0-0.9997).TotalMilliseconds = 0
if(millisecondsTillNextPoll < s_sleepDurationInMilliseconds && millisecondsTillNextPoll > 0) evaluates as if(true && false), therefore s_sleepDurationInMilliseconds = 0 doesn't happen. 
Sleep(999) runs again.
OnTimer() gets run late at ~1.9987s

I'd suggest setting millisecondsTillNextPoll to a minimum value of 1 and calling OnTimer() if nextPollingTimeStamp < now + 1ms.